<launch>
  <!-- Launch Parameters -->
  <arg name="cameraName" value="camera" />
  <arg name="targetName" value="ugv0" />
  <arg name="recordData" value="false" />
  <arg name="showViz" value="false" />
  <arg name="ekf" value="true" />
  <arg name="estimator" value="true" />
  <arg name="predictor" default="true" /> 
  <arg name="normalizedKinematics" value="true" />
  <arg name="deadReckoning" default="true" /> <!-- use turtlebot onboard odometry for rotation estimation -->
  <arg name="artificialSwitching" default="true" /> <!-- artificial switching (i.e. fake loss of visibility) -->
  <arg name="visibilityTimeout" default="0.1" /> <!-- time to wait before instantiating predictor or zero order hold -->
  <arg name="delTon" default="4.0" /> <!-- contiguous visibility time if artificial switching -->
  <arg name="delToff" default="2.0" /> <!-- contiguous invisibility time if artificial switching -->
  <arg name="X0" default="0.0" /> <!-- Initial condition -->
  <arg name="Y0" default="0.0" /> <!-- Initial condition -->
  <arg name="Z0" default="10.5" /> <!-- Initial condition -->
  <arg name="useVelocityMap" default="true" /> <!-- use velocity map, or communicate target velocities -->
  
  <!-- Simulate measurements -->
  <node name="sim" pkg="switch_vis_exp" type="sim_node" output="screen" >
    <param name="cameraName" value="$(arg cameraName)" />
  </node>
  
  <!-- Control camera and target with xbox controller -->
  <node name="joy" pkg="joy" type="joy_node" output="screen" >
    <param name="deadzone" value="0.15" />
  </node>
  
  <!-- Velocity map -->
  <node pkg="switch_vis_exp" type="velocity_map_node" name="velocity_map" output="screen" />
  
  <!-- Estimator -->
  <node if="$(arg estimator)" name="exp" pkg="switch_vis_exp" type="exp_node" output="screen" >
    <param name="cameraName" value="$(arg cameraName)" />
    <param name="targetName" value="$(arg targetName)" />
    <param name="usePredictor" value="$(arg predictor)" />
    <param name="deadReckoning" value="$(arg deadReckoning)" />
    <param name="artificialSwitching" value="$(arg artificialSwitching)" />
    <param name="useVelocityMap" value="$(arg useVelocityMap)" />
    <param name="visibilityTimeout" value="$(arg visibilityTimeout)" />
    <param name="k1" value="2.0" /> <!-- Gains -->
    <param name="k2" value="2.0" /> <!-- Gains -->
    <param name="k3" value="1.0" /> <!-- Gains -->
    <param name="delTon" value="$(arg delTon)" />
    <param name="delToff" value="$(arg delToff)" />
    <param name="X0" value="$(arg X0)" />
    <param name="Y0" value="$(arg Y0)" />
    <param name="Z0" value="$(arg Z0)" />
    <remap from="output" to="exp/output" />
    <remap from="output_point" to="exp/output_point" />
  </node>
  
  <!-- EKF -->
  <node if="$(arg ekf)" name="ekf" pkg="switch_vis_exp" type="ekf_node" output="screen" >
    <param name="cameraName" value="$(arg cameraName)" />
    <param name="targetName" value="$(arg targetName)" />
    <param name="deadReckoning" value="$(arg deadReckoning)" />
    <param name="normalizedKinematics" value="$(arg normalizedKinematics)" />
    <param name="artificialSwitching" value="$(arg artificialSwitching)" />
    <param name="useVelocityMap" value="$(arg useVelocityMap)" />
    <param name="visibilityTimeout" value="$(arg visibilityTimeout)" />
    <param name="q" value="1.0" />
    <param name="r" value="0.01" />
    <param name="delTon" value="$(arg delTon)" />
    <param name="delToff" value="$(arg delToff)" />
    <param name="X0" value="$(arg X0)" />
    <param name="Y0" value="$(arg Y0)" />
    <param name="Z0" value="$(arg Z0)" />
    <remap from="output" to="ekf/output" />
    <remap from="output_point" to="ekf/output_point" />
  </node>
  
  <!-- Launch Vizualization -->
  <node if="$(arg showViz)" name="rviz" pkg="rviz" type="rviz" />
  
  <!-- Launch Plot -->
  <!-- <node if="$(arg showViz)" name="plot" pkg="rqt_plot" type="rqt_plot" /> -->
  
  <!-- Record data -->
  <node if="$(arg recordData)" name="record" pkg="rosbag" type="record" args="-o /home/ncr/experiment_data/exp_aug_switch.bag /output /debug /filter /markerImage /$(arg cameraName)/camera_info /markerCenters /markers /odom" />
  
</launch>
